timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,954 @@
+import functools
+from collections import defaultdict
+from collections.abc import Iterable
+from enum import Enum
+
+import jax.numpy as jnp
+import numpy as np
+
+from timemachine.fe import dummy, geometry, topology, utils
+from timemachine.fe.geometry import LocalGeometry
+from timemachine.lib import potentials
+from timemachine.potentials import bonded, nonbonded
+
+
+def is_planarizing(force_constant, phase, period):
+    """
+    Determine if this we're trying to enforce planarity of two atoms across a
+    bond. Typically these are encoded by proper torsions.
+    """
+    return period == 2 and abs(phase - np.pi) < 0.05 and force_constant > 10.0
+
+
+def identify_bonds_spanned_by_planar_torsions(proper_idxs, proper_params):
+    """
+    Identify bonds that are spanned by planar torsions and returns a dict of bonds
+    and associated torsions that span it.
+    """
+    planar_bonds = defaultdict(list)
+    torsions = defaultdict(list)
+
+    # collect unique idxs first
+    for idxs, params in zip(proper_idxs, proper_params):
+        idxs = tuple(idxs.tolist())
+        torsions[idxs].append(params)
+
+    for (i, j, k, l), all_params in torsions.items():
+        # planarizing torsions should be unique
+        if len(all_params) == 1:
+            force, phase, period = all_params[0]
+            if is_planarizing(force, phase, period):
+                canon_jk = dummy.canonicalize_bond((j, k))
+                planar_bonds[canon_jk].append((i, j, k, l))
+
+    return planar_bonds
+
+
+def recursive_map(items, mapping):
+    """recursively replace items in a list of list
+    mapping = np.arange(100)[::-1]
+    items = [[0,2,3], [5,1,[2,5,6]], 3]
+    result = recursive_map(items, mapping)
+
+    """
+    if isinstance(items, Iterable):
+        res = []
+        for item in items:
+            res.append(recursive_map(item, mapping))
+        return tuple(res)
```

Comment:
nit: looks like this is a list of tuples, not list of lists

Reply:
will fix the docstring